File: LeetCode/Medium/0031. Next Permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        // algorithm: 
        // start from last, find longest non-increasing suffix
        // set the one before to be the pivot
        // find the rightmost successor to swap with pivot; successor > pivot
        // reverse suffix

        int pivot = nums.length - 1;
        if(pivot == 0) return;
        while(pivot > 0 && nums[pivot] <= nums[pivot-1]) --pivot;
        --pivot;

        if(pivot != -1) {
            int id = nums.length - 1;
            while(id > pivot && nums[pivot] >= nums[id]) --id;
            int tmp = nums[pivot];
            nums[pivot] = nums[id];
            nums[id] = tmp;
        }

        for(int i = pivot+1; i < ((pivot+1)+nums.length)/2; ++i) {
            int tmp = nums[i];
            nums[i] = nums[nums.length + pivot - i];
            nums[nums.length + pivot - i] = tmp;
        }
    }
}

File: LeetCode/Medium/0031. Next Permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.

        approach O(N):
        1. find a breakpoint i from the back where nums[i] < nums[i+1]
        2. find the smallest elem > nums[i] in the right side & swap them
        3. reverse the right side
        """

        for i in range(len(nums)-2, -1, -1):
            if nums[i] < nums[i+1]:
                minVal, minInd = nums[i+1], i+1
                for j in range(i+1, len(nums)):
                    if nums[j] <= minVal and nums[j] > nums[i]:
                        minVal, minInd = nums[j], j
                
                nums[i], nums[minInd] = nums[minInd], nums[i]

                nums[i+1:] = nums[i+1:][::-1]
                return nums
        
        return nums.reverse()
